File: server/tests/test_screenshot_on_demand.js
/**
 * A screenshot an administrator asks for.
 *
 * THE REQUEST. "Agar employee online and working hai to button click and
 * uska current screenshot aa jaye." The schedule already takes pictures at
 * moments nobody chooses; this is the other half — somebody is on a call
 * about what is on that screen right now.
 *
 * WHY IT IS THREE STEPS AND NOT ONE. Nothing here can call the client: it
 * polls /api/config/sync every five seconds. So the click writes a request
 * down, the next sync carries it, and the upload that follows names the
 * request it answers. What this file checks is that those three steps hold
 * together — and, just as much, that they do not pretend:
 *
 *   * a request is only made for somebody who is ONLINE, because a picture
 *     from whenever the app next opened, under a button that says "now",
 *     is hours of difference that nothing on the screen would show;
 *   * handing the request over is not the same as taking the picture, so
 *     the sync does not close it — only an upload does;
 *   * a request nobody answered expires instead of waiting for ever.
 *
 * AND IT IS ON THE RECORD. An administrator looking at somebody's screen on
 * demand is an act they may have to answer for, so it is written where
 * administrative acts are kept, with the name of whoever pressed the button.
 *
 * Run:  node server/tests/test_screenshot_on_demand.js
 */
const { execFileSync } = require("child_process");
const path = require("path");
const { migrate } = require("./_migrate");

const DB = `ets_ssreq_${process.pid}`;
const PORT = 8000 + ((process.pid + 511) % 1000);
const BASE = `http://127.0.0.1:${PORT}/api`;
const PASSWORD = "SuperSecret123";

let failures = 0;
function check(label, ok, detail = "") {
    if (!ok) failures += 1;
    console.log(`  ${ok ? "PASS" : "FAIL"}  ${label}${ok || !detail ? "" : `  — ${detail}`}`);
}

function psql(db, sql) {
    return execFileSync("psql", ["-d", db, "-v", "ON_ERROR_STOP=1", "-tAc", sql],
        { encoding: "utf8" }).trim();
}

async function api(method, route, { token, body } = {}) {
    const response = await fetch(`${BASE}${route}`, {
        method,
        headers: {
            "Content-Type": "application/json",
            ...(token ? { Authorization: `Bearer ${token}` } : {}),
        },
        body: body === undefined ? undefined : JSON.stringify(body),
    });
    let payload = {};
    try { payload = await response.json(); } catch (_) {}
    return { status: response.status, body: payload };
}

/** A capture, uploaded the way the client uploads one. */
async function upload(token, requestId) {
    const form = new FormData();
    form.append("screenshot",
        new Blob([Buffer.from("not-a-real-encrypted-capture")],
            { type: "application/octet-stream" }), "shot.enc");
    if (requestId !== undefined) form.append("request_id", String(requestId));
    const response = await fetch(`${BASE}/screenshots/upload`, {
        method: "POST",
        headers: { Authorization: `Bearer ${token}` },
        body: form,
    });
    let payload = {};
    try { payload = await response.json(); } catch (_) {}
    return { status: response.status, body: payload };
}

const login = async (u, device = "d1") =>
    (await api("POST", "/auth/login",
        { body: { username: u, password: PASSWORD, device_id: device } })).body.token;

const sync = async (token, employee_id, device_id) =>
    api("POST", "/config/sync", { token, body: { employee_id, device_id } });

async function main() {
    const root = path.resolve(__dirname, "..", "..");
    const uploads = require("fs").mkdtempSync(
        path.join(require("os").tmpdir(), "ets_ssreq_"));
    let server;

    try {
        migrate(DB);
        const bcrypt = require(path.join(root, "server", "node_modules", "bcryptjs"));
        const hash = await bcrypt.hash(PASSWORD, 10);
        psql(DB, `INSERT INTO employees (employee_id, username, password, role, full_name) VALUES
            ('SA01','owner','${hash}','super_admin','The Owner'),
            ('A001','admin1','${hash}','admin','Priya Nair'),
            ('E001','rajesh','${hash}','employee','Rajesh Kumar'),
            ('E002','sneha','${hash}','employee','Sneha Iyer')`);

        Object.assign(process.env, {
            DB_HOST: process.env.PGHOST || "127.0.0.1",
            DB_PORT: process.env.PGPORT || "5432",
            DB_NAME: DB,
            DB_USER: process.env.PGUSER || process.env.USER,
            DB_PASSWORD: process.env.PGPASSWORD || "unused-locally",
            JWT_SECRET: "test-secret-not-used-in-production",
            PORT: String(PORT),
            ENCRYPTION_KEY: "0".repeat(64),
            UPLOAD_DIR: uploads,
        });

        ({ server } = require(path.join(root, "server", "server.js")));
        await new Promise((r) => (server.listening ? r() : server.once("listening", r)));

        const admin = await login("admin1", "admin-mac");
        const rajesh = await login("rajesh", "rajesh-laptop");
        // Signed in, and at work: an employee is online when their session is
        // live AND they have a shift open today — presence.js decides this,
        // and this test uses the real answer rather than a second opinion.
        await api("POST", "/attendance/login", { token: rajesh, body: {} });
        const sneha = await login("sneha", "sneha-laptop");

        console.log(`\nWho a screenshot can be asked for (${DB})\n`);

        let res = await api("POST", "/admin/employees/E002/screenshot", { token: admin });
        check("somebody signed in but not at work is refused", res.status === 409,
            `HTTP ${res.status}`);
        check("and told why, by name", /sneha|Sneha/i.test(res.body.message || ""),
            res.body.message);

        res = await api("POST", "/admin/employees/SA01/screenshot", { token: admin });
        check("the owner's screen cannot be asked for — they are not monitored",
            res.status === 403, `HTTP ${res.status}`);

        res = await api("POST", "/admin/employees/E001/screenshot", { token: rajesh });
        check("an employee cannot ask for anybody's screen", res.status === 403,
            `HTTP ${res.status}`);

        res = await api("POST", "/admin/employees/NOBODY/screenshot", { token: admin });
        check("nor can one be asked for somebody who does not exist",
            res.status === 404, `HTTP ${res.status}`);

        console.log("\nThe request, and the poll that collects it");

        res = await api("POST", "/admin/employees/E001/screenshot", { token: admin });
        check("an admin may ask for an employee who is online", res.status === 200,
            `HTTP ${res.status} ${JSON.stringify(res.body).slice(0, 90)}`);
        const requestId = res.body.request_id;
        check("and gets the request back to follow", Boolean(requestId), String(requestId));

        // TWO CLICKS ARE ONE REQUEST. A slow network must not take two
        // pictures a second apart and leave one of them pending for ever.
        res = await api("POST", "/admin/employees/E001/screenshot", { token: admin });
        check("clicking again joins the request already waiting",
            String(res.body.request_id) === String(requestId) && res.body.already_waiting === true,
            `${res.body.request_id} vs ${requestId}`);

        const audit = psql(DB,
            `SELECT activity FROM activity_logs WHERE employee_id='E001'
              AND activity LIKE 'SCREENSHOT REQUESTED%' ORDER BY id DESC LIMIT 1`);
        check("it is on the record, with who asked", audit.includes("A001"), audit || "(nothing)");

        let synced = await sync(rajesh, "E001", "rajesh-laptop");
        check("the employee's next sync carries it",
            String(synced.body.config?.capture_now) === String(requestId),
            String(synced.body.config?.capture_now));

        // HANDED OVER IS NOT TAKEN. The app may be closing; the screen may
        // refuse. Only a picture closes the request.
        check("but the request is still waiting, because no picture exists yet",
            psql(DB, `SELECT status FROM screenshot_requests WHERE id=${requestId}`) === "PENDING",
            psql(DB, `SELECT status FROM screenshot_requests WHERE id=${requestId}`));

        // Nobody else's sync is affected by it.
        const other = await sync(sneha, "E002", "sneha-laptop");
        check("and nobody else is asked for a picture they were not asked for",
            !other.body.config?.capture_now, String(other.body.config?.capture_now));

        console.log("\nThe picture that answers it");

        const sent = await upload(rajesh, requestId);
        check("the capture uploads", sent.status === 200, `HTTP ${sent.status}`);
        const row = psql(DB,
            `SELECT status || '|' || COALESCE(screenshot_id::text,'-')
               FROM screenshot_requests WHERE id=${requestId}`);
        check("the request is closed by it", row.startsWith("TAKEN"), row);
        check("and points at the picture that answered it",
            row.split("|")[1] === psql(DB,
                `SELECT id::text FROM screenshots WHERE employee_id='E001' ORDER BY id DESC LIMIT 1`),
            row);

        synced = await sync(rajesh, "E001", "rajesh-laptop");
        check("the next sync asks for nothing — it is done",
            !synced.body.config?.capture_now, String(synced.body.config?.capture_now));

        res = await api("GET", `/admin/screenshot-requests/${requestId}`, { token: admin });
        check("and the page that was waiting can see it was taken",
            res.body.request?.status === "TAKEN", JSON.stringify(res.body).slice(0, 90));

        console.log("\nA request nobody answers");

        res = await api("POST", "/admin/employees/E001/screenshot", { token: admin });
        const stale = res.body.request_id;
        // The app was closed between the click and the poll.
        psql(DB, `UPDATE screenshot_requests
                     SET requested_at = (NOW() AT TIME ZONE 'UTC') - INTERVAL '6 minutes'
                   WHERE id = ${stale}`);

        synced = await sync(rajesh, "E001", "rajesh-laptop");
        check("a stale request is not handed to the client",
            !synced.body.config?.capture_now, String(synced.body.config?.capture_now));

        res = await api("GET", `/admin/screenshot-requests/${stale}`, { token: admin });
        check("and the page is told it expired rather than waiting for ever",
            res.body.request?.status === "EXPIRED", res.body.request?.status);

        // AND AN UPLOAD THAT NAMES NOTHING IS STILL A SCREENSHOT. The
        // scheduled captures carry no request id, and they must keep working
        // exactly as they did.
        const before = Number(psql(DB, `SELECT COUNT(*) FROM screenshots WHERE employee_id='E001'`));
        const plain = await upload(rajesh);
        check("a scheduled capture still uploads with no request at all",
            plain.status === 200
            && Number(psql(DB, `SELECT COUNT(*) FROM screenshots WHERE employee_id='E001'`)) === before + 1,
            `HTTP ${plain.status}`);
    } finally {
        if (server) server.close();
        try { require(path.join(root, "server", "config", "db")).end(); } catch (_) {}
        execFileSync("psql", ["-d", "postgres", "-c",
            `DROP DATABASE IF EXISTS ${DB} WITH (FORCE)`], { stdio: "pipe" });
        require("fs").rmSync(uploads, { recursive: true, force: true });
    }

    console.log();
    if (failures) {
        console.log(`${failures} failure(s)`);
        process.exit(1);
    }
    console.log("all on-demand screenshot checks passed");
    process.exit(0);
}

main();

File: tests/test_screenshot_on_demand.py
"""
A screenshot an administrator asks for, on the client's side of it.

THE REQUEST. "Agar employee online and working hai to button click and uska
current screenshot aa jaye." Nothing can call this app: it polls the server
every five seconds. So the answer rides back on that poll, the capture names
the request it answers, and the button waits for the picture rather than
announcing one.

WHAT IS CHECKED HERE, and each of these is a way the feature could look like
it works while being useless:

  * the poll notices the request and hands it on — with the id, because a
    capture that cannot say which request it answers leaves the button
    waiting for ever;
  * the capture happens even when the day's budget is spent. The budget
    exists to stop unattended monitoring running away; a person pressing a
    button is not that, and an administrator who exhausted somebody's
    monitoring by asking for pictures would have broken the thing they were
    using;
  * the request id travels with the upload, since that is what closes the
    request;
  * the button is off when the person is offline, and SAYS why — a request
    queued for whenever the app next opened would answer "now" with a
    picture from hours later.

Run:  python3 tests/test_screenshot_on_demand.py
"""

import base64
import json
import os
import sys
import tempfile
import time

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

os.environ.setdefault("ETS_DATA_DIR", tempfile.mkdtemp(prefix="ets_ondemand_"))
os.environ.setdefault("API_BASE_URL", "http://127.0.0.1:9/api")
os.environ.setdefault("QT_QPA_PLATFORM", "offscreen")

_TMP = tempfile.mkdtemp(prefix="ets_ondemand_db_")
import client.core.config as config                                          # noqa: E402
config.STORAGE_DIR = _TMP
from client.infrastructure.database import database as database_module       # noqa: E402
database_module.Database.DB_PATH = os.path.join(_TMP, "ets.db")

failures = 0


def check(label, ok, detail=""):
    global failures
    if not ok:
        failures += 1
    print(f"  {'PASS' if ok else 'FAIL'}  {label}"
          + ("" if ok or not detail else f"  — {detail}"))
    sys.stdout.flush()


from PIL import Image                                                        # noqa: E402
from PySide6.QtWidgets import QApplication                                   # noqa: E402

app = QApplication.instance() or QApplication([])

from client.infrastructure.database.database import Database                 # noqa: E402
from client.application.managers.session_manager import SessionManager       # noqa: E402
from client.services.logger_service import LoggerService                     # noqa: E402

Database.initialize()
SessionManager.employee_id = "E001"
SessionManager.role = "employee"
SessionManager.auth_token = "not-checked-here"

lines = []
LoggerService.log = lambda m, *a, **k: lines.append(str(m))
LoggerService.log_verbose = lambda m, *a, **k: lines.append(str(m))


print("\nThe poll that collects the request")

import client.application.managers.config_sync_manager as sync_mod           # noqa: E402


class _Reply:
    """What the server says when somebody has pressed the button."""

    status_code = 200

    def __init__(self, config):
        self._config = config

    def json(self):
        return {"success": True, "config": self._config}


real_post = sync_mod._http.post
asked = []
try:
    sync_mod._http.post = lambda *a, **k: _Reply({"capture_now": 77,
                                                  "screenshots_per_day": 10})
    manager = sync_mod.ConfigSyncManager(
        employee_id="E001", device_id="d1", auth_token="t",
        on_capture_now=lambda request_id: asked.append(request_id))
    manager.sync_now()
finally:
    sync_mod._http.post = real_post

check("a request on the poll is handed on", asked == [77], str(asked))
check("and it carries the id, not just 'somebody asked'",
      asked and isinstance(asked[0], int), str(asked))

nothing = []
try:
    sync_mod._http.post = lambda *a, **k: _Reply({"screenshots_per_day": 10})
    manager = sync_mod.ConfigSyncManager(
        employee_id="E001", device_id="d1", auth_token="t",
        on_capture_now=lambda request_id: nothing.append(request_id))
    manager.sync_now()
finally:
    sync_mod._http.post = real_post
check("an ordinary poll asks for no picture at all", nothing == [], str(nothing))


print("\nThe scheduler hands it to the panel")

from client.application.schedulers.scheduler_service import SchedulerService  # noqa: E402

scheduler = SchedulerService()
seen = []
scheduler.capture_requested.connect(lambda request_id: seen.append(request_id))
scheduler._handle_capture_now(91)
# It crosses from the sync thread to this one, so it arrives on the loop.
end = time.time() + 2
while time.time() < end and not seen:
    app.processEvents()
    time.sleep(0.02)
check("the request reaches the panel, on the main thread", seen == [91], str(seen))
scheduler.stop()


print("\nThe capture that answers it")

import client.application.managers.screenshot_manager as sm                   # noqa: E402
from client.application.managers.screenshot_manager import ScreenshotManager  # noqa: E402


def rows_now():
    with Database.get_connection() as conn:
        return conn.execute("SELECT COUNT(*) c FROM screenshots").fetchone()["c"]


class _Upload:
    status_code = 200

    def json(self):
        return {"success": True}


posted = []
real_quartz, real_auto, real_http = sm._grab_via_quartz, sm.pyautogui.screenshot, sm._http.post
try:
    sm._grab_via_quartz = lambda: None
    sm.pyautogui.screenshot = lambda *a, **k: Image.new("RGB", (320, 240), (10, 20, 30))
    sm._http.post = lambda url, **kwargs: (posted.append(kwargs), _Upload())[1]

    # THE DAY'S BUDGET IS SPENT. A scheduled capture would be refused here.
    ScreenshotManager.screenshots_per_day = classmethod(lambda cls: 1)
    ScreenshotManager.captures_today = classmethod(lambda cls: 5)

    lines.clear()
    before = rows_now()
    refused = ScreenshotManager.capture_screenshot()
    check("a scheduled capture stops when the budget is spent",
          refused is None and any("SKIPPED" in m for m in lines), str(lines)[:120])

    lines.clear()
    posted.clear()
    taken = ScreenshotManager.capture_screenshot(request_id=77)
    check("but one that was asked for is taken anyway",
          taken is not None, str(lines)[:160])
    check("and is recorded as asked for, not as a scheduled one",
          any("ON REQUEST" in m for m in lines), str(lines)[:160])
    check("it is stored like any other", rows_now() == before + 1,
          f"{before} -> {rows_now()}")
    # WITHOUT THIS THE BUTTON WAITS FOR EVER. The server closes the request
    # on the upload that names it.
    sent = posted[-1] if posted else {}
    check("and the upload says which request it answers",
          (sent.get("data") or {}).get("request_id") == "77",
          str(sent.get("data")))
    # AND A SCHEDULED ONE CARRIES NOTHING. Sending an id with every capture
    # would close whatever request happened to be open with a picture nobody
    # asked for at that moment.
    ScreenshotManager.captures_today = classmethod(lambda cls: 0)
    ScreenshotManager.screenshots_per_day = classmethod(lambda cls: 10)
    posted.clear()
    ScreenshotManager.capture_screenshot()
    check("while a scheduled one says nothing about requests",
          posted and not (posted[-1].get("data") or {}),
          str(posted[-1].get("data") if posted else "nothing uploaded"))
finally:
    sm._grab_via_quartz, sm.pyautogui.screenshot = real_quartz, real_auto
    sm._http.post = real_http
    del ScreenshotManager.screenshots_per_day
    del ScreenshotManager.captures_today


print("\nThe button an administrator presses")

from client.presentation.windows import admin_config_panel as panel          # noqa: E402

posts = []


class _CaptureFetch:
    def __init__(self, url, params=None, *a, **k):
        posts.append(("GET", url))

    def __getattr__(self, _name):
        return type("_Sig", (), {"connect": lambda *_a, **_k: None})()

    def start(self):
        pass


class _CapturePost:
    def __init__(self, url, body=None, *a, **k):
        posts.append(("POST", url))

    def __getattr__(self, _name):
        return type("_Sig", (), {"connect": lambda *_a, **_k: None})()

    def start(self):
        pass


real_fetch, real_post_worker, real_track = (
    panel._FetchWorker, panel._PostWorker, panel._track_worker)
try:
    panel._FetchWorker = _CaptureFetch
    panel._PostWorker = _CapturePost
    panel._track_worker = lambda *a, **k: None

    page = panel.EmployeePage()
    page.load({"employee_id": "E001", "username": "rajesh",
               "full_name": "Rajesh Kumar", "role": "employee"})
    check("the button is off until we know they are online",
          not page._shot_now.isEnabled())
    check("and says that is why, rather than just being grey",
          "not online" in page._shot_now.toolTip(), page._shot_now.toolTip())

    page._employee_online = True
    page._update_capture_button()
    check("online, it can be pressed", page._shot_now.isEnabled())
    check("and says what pressing it does, and that it is recorded",
          "right now" in page._shot_now.toolTip()
          and "audit" in page._shot_now.toolTip(), page._shot_now.toolTip())

    posts.clear()
    page._request_screenshot()
    check("pressing it asks the server for this employee",
          posts and posts[-1] == ("POST", f"{panel.API_BASE_URL}/admin/employees/E001/screenshot"),
          str(posts[-1] if posts else "nothing"))

    page._capture_asked({"success": True, "request_id": 5})
    check("then it waits rather than claiming a picture exists",
          not page._shot_now.isEnabled() and "Wait" in page._shot_now.text(),
          page._shot_now.text())
    check("and the page polls for the answer",
          page._capture_timer.isActive())

    page._capture_polled({"request": {"status": "PENDING"}})
    check("a request still pending keeps it waiting",
          page._capture_timer.isActive() and not page._shot_now.isEnabled())

    real_info = panel.QMessageBox.information
    panel.QMessageBox.information = staticmethod(lambda *a, **k: None)
    try:
        page._capture_polled({"request": {"status": "TAKEN", "screenshot_id": 9}})
    finally:
        panel.QMessageBox.information = real_info
    check("a picture ends the wait and frees the button",
          page._shot_now.isEnabled() and not page._capture_timer.isActive()
          and page._shot_now.text() == "Screenshot now", page._shot_now.text())

    page.deleteLater()
    app.processEvents()
finally:
    panel._FetchWorker, panel._PostWorker, panel._track_worker = (
        real_fetch, real_post_worker, real_track)

print(f"\n{'ALL PASS' if not failures else str(failures) + ' FAILED'}\n")

sys.stdout.flush()
sys.stderr.flush()
os._exit(1 if failures else 0)
